accepted/file-enumeration.md

```diff
@@ -0,0 +1,238 @@
+			FindPredicate<TState> recursePredicate = null,
+            TState state = default,
+            FindOptions options = FindOptions.None);
+    }
+
+    /// <summary>
+    /// Used for processing and filtering find results.
+    /// </summary>
+    public ref struct RawFindData
+    {
+        // This will have private members that hold the native data and
+        // will lazily fill in data for properties where such data is not
+        // immediately available in the current platform's native results.
+
+		// The full path to the directory the current result is in
+        public string Directory { get; }
+
+        // The full path to the starting directory for enumeration
+        public string OriginalDirectory { get; }
+
+        // The path to the starting directory as passed to the enumerable constructor
+        public string OriginalUserDirectory { get; }
+
+        // Note: using a span allows us to reduce unneeded allocations
+        public ReadOnlySpan<char> FileName { get; }
+        public FileAttributes Attributes { get; }
+        public long Length { get; }
+
+        public DateTime CreationTimeUtc { get; }
+        public DateTime LastAccessTimeUtc { get; }
+        public DateTime LastWriteTimeUtc { get; }
+    }
+}
+```
+
+### Transforms & Predicates
+
+We'll provide common predicates transforms for building searches.
+
+``` C#
+namespace System.IO
+{
+    internal static partial class FindPredicates
+    {
+        internal static bool NotDotOrDotDot(ref RawFindData findData)
+        internal static bool IsDirectory(ref RawFindData findData)
+    }
+
+    public static partial class FindTransforms
+    {
+        public static DirectoryInfo AsDirectoryInfo(ref RawFindData findData)
+        public static FileInfo AsFileInfo(ref RawFindData findData)
+        public static FileSystemInfo AsFileSystemInfo(ref RawFindData findData)
+        public static string AsFileName(ref RawFindData findData)
+        public static string AsFullPath(ref RawFindData findData)
+    }
+}
+
+```
+
+### DosMatcher
+
+We currently have an implementation of the algorithm used for matching files on Windows in FileSystemWatcher. Providing this publicly will allow consistently matching names cross platform according to Windows rules if such behavior is desired.
+
+``` C#
+namespace System.IO
+{
+    public static class DosMatcher
+    {
+        /// <summary>
+        /// Change '*' and '?' to '&lt;', '&gt;' and '"' to match Win32 behavior. For compatibility, Windows
+        /// changes some wildcards to provide a closer match to historical DOS 8.3 filename matching.
+        /// </summary>
+        public unsafe static string TranslateExpression(string expression)
+
+        /// <summary>
+        /// Return true if the given expression matches the given name.
+        /// </summary>
+        public unsafe static bool MatchPattern(string expression, ReadOnlySpan<char> name, bool ignoreCase = true)
+    }
+}
+```
+
+### Samples
```

Comment:
>Samples [](start = 4, length = 7)

I'd take both the `DosMatcher` as well as this sample and make it part of a scenario.

Reply:
Ok

---
In reply to: [154807983](https://github.com/dotnet/designs/pull/24#discussion_r154807983) [](ancestors = 154807983)